### PR TITLE
RFC 0001: async macro should rewrite return type to Future[T]

### PR DESCRIPTION
**Pull request by zevv**

### CODE REVIEW

**Comment by GULPF** [comment]
I don't like allowing an unwrapped type since it's not the true return type. If I call `proc() {.async.}: Future[int]` I wont receive an `int`, I'll receive a `Future[int]` which must be unpacked using `await`.

**Comment by zah** [comment]
I think we should have had this implicit rewrite, since omitting `Future` is always an error.

It's one of those things that I often forget only to see the compilation failing a moment later - it's few seconds of lost productivity paid from time to time.

**Comment by zevv** [comment]
I think we are making the same point, but from a different perspective. I don't like the explicit ``Future[T]`` return type because it is not the true type of ``result``: the proc signature tells me the return type is a ``Future[int]``, but I have to return an ``int`` instead. The ``async`` macro is changing the interface of the function, and it can either do it on the outside (return type) or on the inside (type of ``result``). 

My goal of this RFC is to have the least amount of friction while writing code. I believe this change is for the good.

And isn't ``await`` just the typical use case for calling async functions?

File: RFCs/0001-Implicit-Future-return-type-transformation-by-async-macro.md [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+## Abstract
+
+The ``async`` macro should automatically transform the return type ``T`` of the
+procedure into ``Future[T]``
+
+
+## Motivation
+
+The current implementation of async functions require explicit ``Future[T]``
+return types. While this might add more clarity, this is also confusing and
```
[the remark this answers is not in the document]
Comment (zevv) [reply RFCs/0001-Implicit-Future-return-type-transformation-by-async-macro.md]: Yeah, the "more clarity" was quoted from you on #Nim, but I found it confusing :)

**Comment by dom96** [comment]
We've discussed this already in a previous issue/RFC (if someone can find it then please do).

Thank you for writing a "proper" RFC for this, you don't see those often (and I wish we did). Sadly I am opposed to this completely, for multiple reasons.

> The current implementation of async functions require explicit ``Future[T]``
return types. While this might add more clarity, this is also confusing and

IMO not including `Future` in the return type will be more confusing. 

Another reason is that other languages do not omit this, C# for example uses ``Task<int>`` and allows you to return a int. This is another reason why this change would be more confusing, this feature is modelled and inspired by C# and diverging from it in this way is a mistake.

I don't believe the burden of writing the return type explicitly warrants this change, especially since this is incredibly centric to my book and to a lot of code out there.

**Comment by cheatfate** [comment]
Explicit `Future[T]` is needed until more people will join Nim's async programming world and will be able to do things in proper way.
I really do not like Nim's way of doing one thing in many ways, e.g. `addr a`, `addr(a)`, `a.addr`, so lets not make the same problem in async.

**Comment by dom96** [comment]
Okay, I'm going to close this.

**Pull request closed without merging**